taichi/backends/opengl/opengl_api.cpp

```diff
@@ -340,25 +340,35 @@ void CompiledProgram::set_used(const UsedFeature &used) {
 OpenGlRuntime::~OpenGlRuntime() = default;
 
 void DeviceCompiledProgram::launch(RuntimeContext &ctx,
+                                   Kernel *kernel,
                                    OpenGlRuntime *runtime) const {
   uint8_t *args_buf_mapped = nullptr;
+  auto args = kernel->args;
 
   // Prepare external arrays/ndarrays
-  // Each external array is copied into its own device buffer (SSBO).
-  // - ctx.args[i] contains its ptr on host.
-  // - arr_bufs_[i] contains its DeviceAllocation on device.
+  // - ctx.args[i] contains its ptr on host, it could be a raw ptr or
+  // DeviceAllocation*
+  // - For raw ptrs, arr_bufs_[i] contains its DeviceAllocation on device.
   // Note shapes of these external arrays still reside in argument buffer,
   // see more details below.
-  for (const auto &[i, size] : program_.ext_arr_map) {
-    if (size == 0)
+  for (int i = 0; i < args.size(); i++) {
+    if (!args[i].is_external_array)
       continue;
-    void *baseptr = device_->map(arr_bufs_[i]);
-    if (program_.check_ext_arr_read(i)) {
-      std::memcpy((char *)baseptr, (void *)ctx.args[i], size);
+    if (args[i].size == 0)
+      continue;
+
+    if (!ctx.is_device_allocation[i]) {
+      void *host_ptr = (void *)ctx.args[i];
+      const_cast<DeviceAllocation &>(arr_bufs_[i]) = device_->allocate_memory(
```

Comment:
Hmm, this leaks SSBO memory. I believe the correct logic is:

1. If `arr_bufs_[i]` is empty, allocate
2. If `args[i].size` doesn't match the current `arr_bufs_[i]`, deallocate the existing `arr_bufs_[i]`, then allocate a new one
3. Deallocate in the destructor.

Reply:
ohhhh nice catch! Thanks a lot!